background-charm-service/src/space-manager.ts

```diff
@@ -169,40 +180,107 @@ export class SpaceManager {
 
     try {
       await this.workerController.runCharm(bg);
-      this.recordSuccess(charmId, bg);
+      this.onProcessSuccess(charmId, bg);
     } catch (error) {
       const errorString = error instanceof Error
         ? error.message
         : String(error);
       log(`${this.did} ${errorString}`, {
         error: true,
       });
-      this.disableCharm(charmId, bg, errorString);
+      this.onProcessFail(charmId, bg, errorString);
     }
     this.activeBg = null;
   }
 
-  private recordSuccess(charmId: string, bg: Cell<BGCharmEntry>) {
+  private onProcessSuccess(charmId: string, bg: Cell<BGCharmEntry>) {
+    // If previous runs have failed, clear out the counter
+    if (this.failureTracking.has(charmId)) {
+      this.failureTracking.delete(charmId);
+    }
+
     bg.update({
       lastRun: Date.now(),
       status: "Success",
     });
+
     if (this.schedulableBgs.has(charmId)) {
-      this.addPendingRun(charmId, bg, this.rerunIntervalMs / 1000);
+      this.addPendingRun(charmId, bg);
+    }
+  }
+
+  private onProcessFail(
+    charmId: string,
+    bg: Cell<BGCharmEntry>,
+    error: string,
+  ) {
+    const failureCount = (this.failureTracking.get(charmId) ?? 0) + 1;
+
+    // If we've received graph errors 3 times in a row,
+    // disable the charm.
+    if (failureCount >= 3) {
+      this.failureTracking.delete(charmId);
+      this.disableCharm(charmId, bg, error);
+    } else {
+      this.failureTracking.set(charmId, failureCount);
+      bg.update({
+        lastRun: Date.now(),
+        status: error,
+      });
+      if (this.schedulableBgs.has(charmId)) {
+        this.addPendingRun(charmId, bg);
```

Comment:
I wonder if we should do some sort of exponential backoff as well?

Reply:
SG, will apply a linear backoff here